src/utils/spark.py

```diff
@@ -0,0 +1,223 @@
+        self.final_dir = (
+            (Path(final_dir) / strip_table_prefix(self.table_name))
+            .resolve()
+            .as_posix()
+        )
+
+        # Both of these must be set to avoid situations where we create
+        # partitions by taxyr but not cur, and visa-versa
+        if self.use_partitions and (
+            (self.taxyr is None) != (self.cur is None)
+        ):
+            raise ValueError(
+                (
+                    f"Error for table {self.table_name}: "
+                    "both 'taxyr' and 'cur' must be set if one is set."
+                )
+            )
+
+    def get_filter(self) -> str | None:
+        """
+        Translates the taxyr and cur values into SQL used to filter/limit the
+        values read from the table.
+        """
+        if self.taxyr is None:
+            return None
+
+        if isinstance(self.taxyr, list):
+            filter = f"taxyr IN ({', '.join(map(str, self.taxyr))})"
+        else:
+            filter = f"taxyr = {self.taxyr}"
+
+        if isinstance(self.cur, list):
+            quoted_cur = [f"'{x}'" for x in self.cur]
+            filter += f" AND cur IN ({', '.join(quoted_cur)})"
+        else:
+            filter += f" AND cur = '{self.cur}'"
+        return filter
+
+    def get_partition(self) -> list[str]:
+        """
+        Partition values used for Hive-style partitions e.g. the outputs from
+        read() will look something like:
+
+            /tmp/target/initial/addn/taxyr=2020/cur=Y/big_file_name1.parquet
+            /tmp/target/initial/addn/taxyr=2021/cur=N/big_file_name1.parquet
+        """
+        return ["taxyr", "cur"] if self.use_partitions else []
+
+    def read(self) -> None:
+        """
+        Perform the JDBC read and the initial file write to disk. Files will be
+        partitioned by the number of values passed via predicates (by default
+        96 per taxyr).
+        """
+        filter = self.get_filter()
+        partitions = self.get_partition()
+
+        # Must use the JDBC read method here since the normal spark.read()
+        # doesn't accept predicates https://stackoverflow.com/a/48680140
+        df = self.session.spark.read.jdbc(
+            url=self.session.database_url,
+            table=self.table_name,
+            predicates=self.predicates,
+            properties={
+                "user": self.session.ipts_username,
+                "password": self.session.ipts_password,
+                "fetchsize": self.session.fetch_size,
+            },
+        )
+
+        # Only apply the filtering step if limiting values are actually passed
+        # because it errors with an empty string or None value
+        if filter:
+            df = df.filter(filter)
+
+        (
+            df.write.mode("overwrite")
+            .option("compression", self.session.initial_compression)
+            .partitionBy(partitions)
+            .parquet(self.initial_dir)
+        )
+
+    def repartition(self) -> None:
```

Comment:
This method is called after _all_ the Spark extract jobs have run. If you try to call it after each job it will crash the server. Spark doesn't like to free memory, neither does PyArrow. They fight to the death and chungus loses.